ui: add a simple proxy server

maxlang opened the pull request:
This proxy server allows a developer to point a local Cockroach UI
to an existing Cockroach instance.

Made with the assistance of @mrtracy

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9858)

<!-- Reviewable:end -->

Commit:
ui: add a simple proxy server

This proxy server allows a developer to point a local Cockroach UI
to an existing Cockroach instance.

mrtracy commented:
I would suggest moving this to a separate directory; suggesting `ui/proxy`. It could have a separate npm installation under that directory; that would separate these extra dependencies, which are not necessary to build the actual web UI.

---

Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[ui/proxy.js, line 12 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9858#-KTk9g-ru40CtTCv7NVv:-KTk9g-saR0x3IRuFKun:b-lamlzf) ([raw file](https://github.com/cockroachdb/cockroach/blob/70ee90ced9fa061c05b63d5995dd70c427c81b56/ui/proxy.js#L12)):*
> ```JavaScript
> 
> var argv = require('yargs')
>   .usage('Usage: $0 <remote-cockroach-ui-port> [options]')
> ```

This is a remote URL, not a remote port.  s/remote-cockroach-ui-port/remote-cockroach-ui-url/g

---

*[ui/proxy.js, line 14 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9858#-KTkA6zT52Rf20bGDG1a:-KTkA6zT52Rf20bGDG1b:b-3jt2ir) ([raw file](https://github.com/cockroachdb/cockroach/blob/70ee90ced9fa061c05b63d5995dd70c427c81b56/ui/proxy.js#L14)):*
> ```JavaScript
>   .usage('Usage: $0 <remote-cockroach-ui-port> [options]')
>   .demand(1)
>   .default('local', 'http://localhost:8080', 'Cockroach instance ui port to serve UI files from')
> ```

This is also a URL, not a port.

---

*[ui/proxy.js, line 16 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9858#-KTk9nyNbj6Ekxt0MyxH:-KTk9nyNbj6Ekxt0MyxI:b-64m2mn) ([raw file](https://github.com/cockroachdb/cockroach/blob/70ee90ced9fa061c05b63d5995dd70c427c81b56/ui/proxy.js#L16)):*
> ```JavaScript
>   .default('local', 'http://localhost:8080', 'Cockroach instance ui port to serve UI files from')
>   .default('port', 3000, 'The port to run this proxy server')
>   .example(`$0 https://myroach:8080`, 'Serve the UI from localhost:8080 and proxy requests to myroach:8080')
> ```

"Serve UI resources (HTML, JS, CSS) from localhost:8080, while serving API data requests from https://myroach:8080"

---

*[ui/README.md, line 59 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9858#-KTk7myBQdW_GSpkGT14:-KTk7myBQdW_GSpkGT15:b-33x51z) ([raw file](https://github.com/cockroachdb/cockroach/blob/70ee90ced9fa061c05b63d5995dd70c427c81b56/ui/README.md#L59)):*
> ```Markdown
> ## Proxying
> 
> To test UI changes against an existing CockroachDB instance, you can use the
> ```

Maybe a little more detail in this section, including a bit of information about why you would want it.

"When prototyping changes to the CockroachDB Admin UI, it is desirable to see those changes with data from an existing cluster *without* the headache of having to redeploy a cluster. This is useful for rapidly visualizing local development changes against a consistent and realistic dataset.

We have created a simple NodeJS reverse-proxy server to accomplish this; this server proxies all requests for web resources (javascript, HTML, CSS) to a local CockroachDB server, while proxying all requests for actual data to a remote CockroachDB server.

To activate this server, run:
`./proxy.js <existing-instance-ui-url> --local <development-instance-ui-url>`

and navigate to "http://localhost:3000" to access the UI.

Run `./proxy.js --help` for detailed information.
"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9858#-:-KTkAx7KSHpPzyxDJkyR:b-cjnqu9)*
<!-- Sent from Reviewable.io -->

tamird commented on ui/README.md:
```diff
@@ -54,6 +54,12 @@ The UI also supports live reload in debug mode. To take advantage of this, run
+
+To test UI changes against an existing CockroachDB instance, you can use the
+`proxy.js` utility. Simply run
+`./proxy.js <existing-instance-ui-port> --local <development-instance-ui-port>`.
```
this is not accurate, AFAICT

maxlang commented:
Added some catchall error handling based on https://github.com/nodejitsu/node-http-proxy/blob/master/examples/http/custom-proxy-error.js

Moved into `proxy` folder as suggested and added a mention in the README.

---

Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[ui/README.md, line 59 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9858#-KTk7myBQdW_GSpkGT14:-KUJMTH8axMeLsEY9QTD:b-wlvu73) ([raw file](https://github.com/cockroachdb/cockroach/blob/70ee90ced9fa061c05b63d5995dd70c427c81b56/ui/README.md#L59)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> Maybe a little more detail in this section, including a bit of information about why you would want it.
> 
> "When prototyping changes to the CockroachDB Admin UI, it is desirable to see those changes with data from an existing cluster *without* the headache of having to redeploy a cluster. This is useful for rapidly visualizing local development changes against a consistent and realistic dataset.
> 
> We have created a simple NodeJS reverse-proxy server to accomplish this; this server proxies all requests for web resources (javascript, HTML, CSS) to a local CockroachDB server, while proxying all requests for actual data to a remote CockroachDB server.
> 
> To activate this server, run:
> `./proxy.js <existing-instance-ui-url> --local <development-instance-ui-url>`
> 
> and navigate to "http://localhost:3000" to access the UI.
> 
> Run `./proxy.js --help` for detailed information.
> "
</details>
Done. Just used your text wholesale with a note about running `npm install`

---

*[ui/README.md, line 61 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9858#-KTk7_PzB6bli_I9r3dl-r1-61:-KUJMUSlJol_xAV980Dp:bfnir9x) ([raw file](https://github.com/cockroachdb/cockroach/blob/70ee90ced9fa061c05b63d5995dd70c427c81b56/ui/README.md#L61)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this is not accurate, AFAICT
</details>
Do you mean the port bit? Should be fixed with Matt's changes.

---

*[ui/proxy.js, line 12 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9858#-KTk9g-ru40CtTCv7NVv:-KUJGqqClBoj8z3ONzJN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/70ee90ced9fa061c05b63d5995dd70c427c81b56/ui/proxy.js#L12)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> This is a remote URL, not a remote port.  s/remote-cockroach-ui-port/remote-cockroach-ui-url/g
</details>
Done.

---

*[ui/proxy.js, line 14 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9858#-KTkA6zT52Rf20bGDG1a:-KUJH-a3vKe59uM2DPHC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/70ee90ced9fa061c05b63d5995dd70c427c81b56/ui/proxy.js#L14)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> This is also a URL, not a port.
</details>
Done.

---

*[ui/proxy.js, line 16 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9858#-KTk9nyNbj6Ekxt0MyxH:-KUJHDbrpncnEpoXY4kM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/70ee90ced9fa061c05b63d5995dd70c427c81b56/ui/proxy.js#L16)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> "Serve UI resources (HTML, JS, CSS) from localhost:8080, while serving API data requests from https://myroach:8080"
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9858#-:-KUJFB4lx-jBwOGge7PH:bsnnqt7)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with one suggestion.

---

Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/ui/proxy/proxy.js, line 54 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9858#-KUJYFcbI7X4p2OMvBm5:-KUJYFccqqguE0Eu2XKo:b27u6gk) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b7ae46f76fc0d0594f35e1fe8f5a502e00ee198/pkg/ui/proxy/proxy.js#L54)):*
> ```JavaScript
>   });
> 
>   res.end('Something went wrong. And we are reporting a custom error message.');
> ```

No need for the "reporting a custom error message", that just looks copy-pasted. Could you print the error text instead of this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9858#-:-KUJYbXywBRprZH4M3lP:btrovwl)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/ui/proxy/proxy.js, line 54 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9858#-KUJYFcbI7X4p2OMvBm5:-KUJZBTqhYgRyR-NLbvB:b-wzdx8v) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b7ae46f76fc0d0594f35e1fe8f5a502e00ee198/pkg/ui/proxy/proxy.js#L54)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> No need for the "reporting a custom error message", that just looks copy-pasted. Could you print the error text instead of this?
</details>
Haha whoops, very copy pasted. Was messing around with error handling and forgot to clean up.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9858)*
<!-- Sent from Reviewable.io -->

maxlang commented:
@tamird merged because I figured I'd addressed your feedback, but let me know if you have any remaining comments.

tamird commented:
Reviewed 3 of 7 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9858)*
<!-- Sent from Reviewable.io -->
